### ISSUE TITLE
[HH] Kimi K3 full-MXFP4 DSpark DCP8/DCP16 status

### ISSUE THREAD

**Issue by voipmonitor** [local-inference-lab/vllm#239]
## TL;DR

| Use case | Recommended profile | Measured result |
|---|---|---:|
| Maximum validated CC1 throughput | DCP16 + BF16 DSpark + selective KDA-input MXFP8 | about 106-115 tok/s; physical target KV 1,057,049 |
| Exact target checkpoint, no speculation, fastest CC1 | DCP8 no-DSpark | 52.826 tok/s matched stable median; 52.358 tok/s immutable-image stable median; physical KV 1,054,602 |
| Exact target checkpoint, extra physical KV / CC8 | DCP16 no-DSpark | 51.985 tok/s matched stable median; physical KV 1,460,937; previously 204.559 tok/s aggregate CC8 |

For exact-target single-request decode, DCP8 is now the default. In the matched
back-to-back A/B it is 1.62% faster than equivalently patched DCP16 and 9.23%
faster than DCP8 r4. Routed experts remain checkpoint-native MXFP4, dense
weights remain BF16, KV is FP8, and the optimization changes no model
arithmetic or reduction order.

Current reproducible runtime:

- vLLM [#238](https://github.com/local-inference-lab/vllm/pull/238), open and non-draft: `3846d740fcfe566e821c574892574f7797b85008`
- SparkInfer [#118](https://github.com/local-inference-lab/sparkinfer/pull/118), open and non-draft: `5a46e5b5a8a87012a5b8261b81130ee057591d5d`
- SparkInfer [#116](https://github.com/local-inference-lab/sparkinfer/pull/116) is the merged FP8-gather/native-E8M0/DCP graph-safety prerequisite; #118 is additive
- Docker image: `voipmonitor/vllm:kimi-k3-hh-runtime-pr238-pr118-r5-20260805@sha256:d5b7e0160ebdb8202237d378bfddd2343fcb4dbd3b537e026f55bec122bf2181`
- [Docker recipe and runner](https://github.com/local-inference-lab/blackwell-llm-docker/tree/build/kimi-k3-hh-runtime-20260805): docs/runner head `fd8181e`, image-build revision `c95180778127cbcb1c4c15fb3ffd52593bf67d42`
- [Runtime matrix and exact launch commands](https://github.com/local-inference-lab/rtx6kpro/blob/docs/kimi-k3-runtime-matrix-20260805/models/kimi-k3/hh-runtime-matrix-2026-08-05.md): docs head `7b32e88`

## Start a profile

The host Hugging Face cache must contain
`moonshotai/Kimi-K3@2496450e92e425c886db095102a52a6682ca3970`. DSpark
profiles also need
`Inferact/Kimi-K3-DSpark@cf6b8244620e7ea4b0651d214f28e89eac75bed6`.

```bash
git clone --branch build/kimi-k3-hh-runtime-20260805 \
  https://github.com/local-inference-lab/blackwell-llm-docker.git \
  blackwell-llm-docker-kimi-k3
cd blackwell-llm-docker-kimi-k3

./run-kimi-k3-hh.sh dcp8-no-dspark
./run-kimi-k3-hh.sh dcp16-no-dspark
./run-kimi-k3-hh.sh dcp16-no-dspark-batch8
./run-kimi-k3-hh.sh dcp16-dspark
./run-kimi-k3-hh.sh dcp16-dspark-full
./run-kimi-k3-hh.sh dcp8-dspark
```

The runner publishes host port 8000, mounts `/root/.cache/huggingface`, enables
all 16 GPUs, and sets
`PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True`. Weight loading is
`--load-format instanttensor`.

## Runtime matrix

| Profile | Target changes | Scheduler | Physical target KV | Decode |
|---|---|---:|---:|---:|
| DCP8 no-DSpark | none; MXFP4 experts/BF16 dense | 256 | 1,054,602 | 52.826 tok/s matched stable; 52.358 immutable-image stable |
| DCP16 no-DSpark | none; MXFP4 experts/BF16 dense | 2048 | 1,460,937 | 51.985 tok/s matched stable |
| DCP16 DSpark/KDA-MXFP8 | 69 KDA input projections MXFP8; routed experts MXFP4; shared experts BF16 | 4096 | 1,057,049 | about 106-115 tok/s CC1 |
| DCP16 DSpark full-target | no target online quantization; BF16 draft | 2048 | physical 1M | functional reference profile |
| DCP8 DSpark legacy | target shared experts MXFP8; BF16 draft | 2048 | 1,048,576 | 99.65 tok/s CC1 median |

## DCP8 inversion: cause and fix

There were two independent DCP8 penalties.

1. Kimi projections are TP16-sharded, but the first B12X projection
   integration selected the DCP coordinator and required
   `DCP world_size == TP size`. DCP8 silently used generic NCCL gathers and
   separate top-k. The earlier PR #238 fix keeps MLA communication on DCP8
   while projection traffic uses the full TP16 SparkInfer pool.
2. After that fix, DCP8 collectives were individually faster but end-to-end
   decode still trailed DCP16. The r4 Torch traces showed that DCP8's shorter
   query gather exposed eager host work before dense MLA: 39.7 us/layer from
   gather completion to MLA launch versus 23.1 us/layer for DCP16. The 16.6
   us/layer difference is about 0.40 ms/token over 24 full-attention layers. A
   direct SparkInfer measurement found `dense_mla.bind` cost 28.47 us/call, or
   0.683 ms over those layers.

vLLM commit `e5865bdc3f` fixes the second penalty:

- metadata owns capture-stable gathered-query and MLA-output arenas
- DCP query gather writes directly into the stable query address
- each full-attention layer reuses its validated SparkInfer binding
- a changed plan, tensor pointer/layout, scale, or active split rebuilds the binding
- fallback paths retain exact behavior

No weight, KV format, precision, active-split policy, arithmetic, or reduction
order changes.

Matched full-model benchmark: fixed 256-token prompt, token ID 13 forced at
every step, 512-token warmup, six 1,024-token decode runs per profile. Decode
timing excludes prefill and the first output token.

| Runtime | Six runs (tok/s) | All-run median | Stable last-three median |
|---|---|---:|---:|
| DCP8 r4 | 46.635, 47.117, 47.849, 48.631, 48.364, 48.162 | 48.006 | 48.364 |
| DCP16 r4 | 47.748, 48.470, 49.110, 49.342, 49.196, 49.419 | 49.153 | 49.342 |
| DCP8 binding-cache candidate | 51.265, 51.289, 52.439, 52.826, 52.782, 52.939 | 52.610 | 52.826 |
| DCP16 binding-cache candidate | 51.798, 51.902, 52.120, 51.985, 51.932, 52.033 | 51.959 | 51.985 |

The candidates used the exact source bytes baked into r5. The immutable r5
image was then started without source mounts and produced a 52.358 tok/s median
in the stable six-run window (range 51.808-52.655). It loaded 497,218 tensors
through InstantTensor in 157 seconds; complete per-rank model loading took
185.9 seconds and 90.96 GiB/rank.

## DCP collective control

The production-geometry SparkInfer CUDA-graph harness confirms that DCP8
itself is faster. Batch 1, BF16, query head width 576, output width 512,
256-thread production launch:

| Geometry | Reduce | Gather | Gather+reduce |
|---|---:|---:|---:|
| DCP8, 48 gathered heads | 14.337 us | 13.303 us | 25.631 us |
| DCP16, 96 gathered heads | 25.170 us | 33.528 us | 57.176 us |

The pair difference is 0.757 ms/token across 24 full-attention layers. DCP8 at
512 threads measured 24.989 us/layer, only 0.015 ms/token better across the
full model, so the result is recorded in the wiki but the tested 256-thread
production default remains.

Rejected model-free experiments:

- a TP16 alignment block before MLA added about 292 us/token
- changing the hierarchical post-barrier sleep away from 24 nanosleep units was worse under injected rank skew
- adaptive 24-to-256 backoff was worse
- the 512-thread DCP8 geometry improved the microbenchmark but is below useful E2E resolution; it is preserved, not deleted

## DSpark results and rejected configurations

| Experiment | Result | Decision |
|---|---|---|
| BF16 draft, DCP1 target | 74.10 tok/s but only 8,894 physical target-KV tokens | reject for context capacity |
| Draft attention forced through target DCP | draft lost the required complete recent context; acceptance collapsed | draft stays DCP1 with bounded replicated tail |
| Whole draft MXFP8 | small-M dynamic W8A8 kernels were 7-14x slower; Marlin W8A16 was faster but acceptance was lower than BF16 | keep draft BF16 |
| Replicated BF16 Markov head | 160 MiB/rank | TP-shard to 10 MiB/rank; 0.353 ms per seven-step block |
| Adaptive proposal depth | target step rate nearly flat for depths 2-8 | keep trained fixed depth seven |
| 64K draft tail | 115.30 tok/s short, 47.39 tok/s at 200K vs 114.97/49.67 with 32K | keep 32K |
| DCP8 BF16 DSpark + sharded Markov | 99.65 tok/s and physical 1M | functional; requires target shared-expert MXFP8, not accuracy default |

The selected DCP16 DSpark profile keeps routed experts MXFP4 and target shared
experts BF16, and converts only 69 low-sensitivity KDA input projections.
Prefill medians at 8K/32K/64K are 3,451/3,589/3,475 tok/s. Decode is about
106-115 tok/s at short context, 110.7 tok/s at 100K, and 49.7 tok/s at 200K.

Selective KDA-input MXFP8 measured
`KL(reference || candidate)=0.003183` over 65,504 positions with 98.826% top-1
agreement. Correct speculative rejection preserves the target distribution;
proposal quality is evaluated separately.

## FlashKDA

FlashKDA is intentionally not in the production PR. The archived SM120 raw-TMA
work reduced persistent first-launch memory from about 3.78 GiB/rank to 0.119
GiB/rank and improved the isolated 65,536-token kernel by about 4.7%, but the
equal-MBT4096 8K end-to-end result was 0.37% slower than Triton. Source, build
changes, harnesses, and measurements are retained in
[`archive/hh-kimi-k3-flashkda-sm120-20260804`](https://github.com/local-inference-lab/vllm/tree/archive/hh-kimi-k3-flashkda-sm120-20260804).

## Validation

- vLLM B12X MLA: 48/48
- vLLM DCP A2A: 48/48 on visible GPUs
- SparkInfer dense MLA: 11/11
- all twelve matched full-model forced-token runs completed
- immutable r5 image SHA and baked vLLM/SparkInfer files match the tested worktrees byte-for-byte
- clean r5 logs show InstantTensor, DCP8 with 48 effective heads, 1,054,602 physical KV tokens, healthy HTTP startup, and no source bind mounts

Rollback before binding reuse:
`voipmonitor/vllm:kimi-k3-hh-runtime-pr238-pr118-r4-20260805@sha256:7e205e5c7c54cb750a480169ee346787e5b09d23351be6e90754ed9440645150`

## Merge checklist

- [ ] Review and merge SparkInfer #118.
- [ ] Review and merge vLLM #238.
- [ ] Rebuild the immutable image from the merged commits.
